File: OLD_USACO/chapter4/section3/buylow.cpp
/* Use the slash-star style comments or the system won't see your
   identification information */
/*
ID: davideo2
TASK: buylow
LANG: C++
*/
/* LANG can be C++11 or C++14 for those more recent releases */
#include <bits/stdc++.h>
using namespace std;

int main() {
    // ifstream cin ("buylow.in");
    // ofstream cout ("buylow.out");
    int n; cin >> n;
    vector<int> arr(n + 1);
    vector<pair<int, int>> dp(n + 1, {1, 1});
    for (int i = 1; i <= n; i++) cin >> arr[i];

    for (int i = 2; i <= n; i++) {
        for (int j = i - 1; j > 0; j--) {
            if (arr[j] <= arr[i]) continue;
            if (dp[j].first + 1 < dp[i].first) continue;
            if (dp[j].first + 1 == dp[i].first) {
                dp[i].second += dp[j].second;
            } else {
                dp[i].first = dp[j].first + 1;
                dp[i].second = dp[j].second;
            }
        }

        for (int j = i - 1; j > 0; j--) {
            if (arr[i] != arr[j]) continue;
            if (dp[i].first != dp[j].first) continue;
            dp[i].second -= dp[j].second;
        }
    }

    int bestLength = 0, numOfSeq = 0;
    for (int i = 1; i <= n; i++) {
        if (dp[i].first > bestLength) {
            bestLength = dp[i].first;
            numOfSeq = dp[i].second;
        } else if (dp[i].first == bestLength) {
            numOfSeq += dp[i].second;
        }
    }

    cout << bestLength << " " << numOfSeq << "\n";
    return 0;
}

File: OLD_USACO/chapter4/section3/buylow.py
"""
ID: davideo2
LANG: PYTHON3
TASK: buylow
"""

def main():
    fin = open('buylow.in', 'r')
    fout = open('buylow.out', 'w')
    
    n = int(fin.readline())
    arr = []
    while (True):
        for num in map(int, fin.readline().split()):
            arr.append(num)
        if (len(arr) == n): break
    
    dp = [[1, 1] for _ in range(n)]

    for i in range(1, n):
        for j in range(i - 1, -1, -1):
            if (arr[j] <= arr[i]): continue
            if (dp[j][0] + 1 < dp[i][0]): continue
            if (dp[j][0] + 1 == dp[i][0]):
                dp[i][1] += dp[j][1]
            else:
                dp[i][0] = dp[j][0] + 1
                dp[i][1] = dp[j][1]

        for j in range(i - 1, -1, -1):
            if (arr[i] != arr[j]): continue
            if (dp[i][0] != dp[j][0]): continue
            dp[i][1] -= dp[j][1]
    
    bestLength = 0
    numOfSeq = 0
    for i in range(n):
        if (dp[i][0] > bestLength):
            bestLength = dp[i][0]
            numOfSeq = dp[i][1]
        elif (dp[i][0] == bestLength):
            numOfSeq += dp[i][1]
    
    fout.write(f"{bestLength} {numOfSeq}\n")

if __name__ == '__main__':
    main()
